pkg/core/interop/context.go

```diff
@@ -150,101 +157,218 @@ type MethodAndPrice struct {
+	// used, there's a mdCache field for that.
+	Events []Event
+	// ActiveHFs is a map of hardforks that contract should react to. Contract update should be called for active
+	// hardforks. Note, that unlike the C# implementation, this map doesn't include contract's activation hardfork.
+	// This map is being initialized on contract creation and used as a read-only, hens, not protected
+	// by mutex.
+	ActiveHFs map[config.Hardfork]struct{}
+
+	// mdCache contains hardfork-specific ready-to-use contract descriptors. This cache is lazy and thus, protected by
+	// mdCacheLock.
+	mdCacheLock sync.RWMutex
+	mdCache     map[config.Hardfork]*HFSpecificContractMD
+
+	// onManifestConstruction is a callback for manifest finalization.
+	onManifestConstruction func(*manifest.Manifest)
+}
+
+// HFSpecificContractMD is a hardfork-specific native contract descriptor.
+type HFSpecificContractMD struct {
+	state.ContractBase
+	Methods []HFSpecificMethodAndPrice
+	Events  []HFSpecificEvent
+}
+
+// NewContractMD returns Contract with the specified fields set. onManifestConstruction callback every time
+// after hardfork-specific manifest creation and aimed to finalize the manifest.
+func NewContractMD(name string, id int32, onManifestConstruction ...func(*manifest.Manifest)) *ContractMD {
 	c := &ContractMD{Name: name}
+	if len(onManifestConstruction) != 0 {
+		c.onManifestConstruction = onManifestConstruction[0]
+	}
 
 	c.ID = id
-
-	// NEF is now stored in the contract state and affects state dump.
-	// Therefore, values are taken from C# node.
-	c.NEF.Header.Compiler = "neo-core-v3.0"
-	c.NEF.Header.Magic = nef.Magic
-	c.NEF.Tokens = []nef.MethodToken{} // avoid `nil` result during JSON marshalling
 	c.Hash = state.CreateNativeContractHash(c.Name)
-	c.Manifest = *manifest.DefaultManifest(name)
+	c.ActiveHFs = make(map[config.Hardfork]struct{})
+	c.mdCache = make(map[config.Hardfork]*HFSpecificContractMD)
 
 	return c
 }
 
-// UpdateHash creates a native contract script and updates hash.
-func (c *ContractMD) UpdateHash() {
+// HFSpecificContractMD returns hardfork-specific native contract metadata, i.e. with methods, events and script
+// corresponding to the specified hardfork. If hardfork is not specified, then default metadata will be returned
+// (methods, events and script that are always active).
+func (c *ContractMD) HFSpecificContractMD(hf *config.Hardfork) *HFSpecificContractMD {
+	var key config.Hardfork
+	if hf != nil {
+		key = *hf
+	}
+	c.mdCacheLock.RLock()
+	if md, ok := c.mdCache[key]; ok {
+		c.mdCacheLock.RUnlock()
+		return md
+	}
+	c.mdCacheLock.RUnlock()
+
+	md := c.buildHFSpecificMD(hf)
+	return md
+}
+
+// buildHFSpecificMD builds hardfork-specific contract descriptor that includes methods and events active starting from
+// the specified hardfork or older.
+func (c *ContractMD) buildHFSpecificMD(hf *config.Hardfork) *HFSpecificContractMD {
```

Comment:
Could it be `contract.MDAtHF()` with keeping HF-specific copies in contracts?

Reply:
No, because we need methods offsets to be resent at this MD.